### ISSUE TITLE
CloudEvent Persistence?

### ISSUE THREAD

**Issue by erikerikson** [cloudevents/spec#276]
As I discussed during community time it seems that there is a certain assumption of transience in this specification and the discussions occurring.  In other words, that relevant concerns are resolved within the context of the connection over which the CloudEvent content is sent and not addressed by CloudEvents.  Concrete examples include authentication as a property of an HTTP connection and event transmission fidelity validated by a TCP checksum.

This issue is opened to facilitate some discussion about the topic.

Certainly some of these concerns should not be solved by CloudEvents and It may be best not to solve for some others given that they comprise a good deal of complexity as well as an ever evolving set of best practices.  That acknowledged, if a CloudEvent is to be written without alteration to a producer and consumer intermediary such as a Kafka or Kinesis the producer's connection context, without event augmentation, is not available to consumers that may have requirements necessitating that information.

A concrete example of this might involve a Kafka stream operated and archived by a banking organization that is used to record, analyze, and actuate inter-account transfers of capital.  While it would be appropriate for me to place an authorization of capital transfer from my account to another, it would be inappropriate for another account holder to have placed the same event on the topic.  Whereas I would be unlikely to directly place events onto such a log, a transfer authorizations service may well do so while a personal information storage service should not.  As an example solution mechanism, a user-specific (and, perhaps, additionally a system-specific) private key signed hash of the CloudEvent could be validated using the associated public key(s) by consumers of the event that required some guarantee of the identity and authorization of the producer.

Obviously any such information could be added to the CloudEvent's [data attribute](https://github.com/cloudevents/spec/blob/master/spec.md#data-attribute) or added as an [extension](https://github.com/cloudevents/spec/blob/master/spec.md#extensions).  However, these needs are present across domains.  Possible alternatives to addressing these concerns within CloudEvents include adding envelopes around persisted CloudEvents or custom solutions within appropriate attribute values of CloudEvents.

Potential questions:
If any, what contextual attributes from connections should be considered for inclusion?
What bases would be helpful in distinguishing between context attributes that should or should not be included?
What adverse outcomes would adding context attributes to the CloudEvents specification cause?

**Comment by rektide** [comment]
For example, Knative Eventing (which is, to my limited understanding, a particular implemention of CloudEvents), has the option to be backed by a Kafka bus. This means events are stored persistently, for some length of time. There's definitely some appeal to me in exposing & making use of persistence in CloudEvent, although I admit I really don't know what that would look like.

**Comment by erikerikson** [comment]
For it's relevance: https://github.com/cloudevents/spec/issues/143#issuecomment-381913100

This comment by Clemens notes some of the challenges of providing auth and signing to the interchange message.

**Comment by duglin** [comment]
@erikerikson what would you like to do with this one? If I'm understanding this correctly, it seems as though there might be multiple issues bundled together into one - and if so, would be useful to split things into individual issues to help focus the discussion?

Do you think this is needed for v0.2?

**Comment by erikerikson** [comment]
There are substantial consequences to the answer on this question.

I doubt that v0.2 should wait on answering this question or resolving the consequences if answered in the affirmative.

**Comment by duglin** [comment]
@erikerikson any thoughts on how to make progress on this one? I still think there are multiple issues in here (and https://github.com/cloudevents/spec/issues/143 might overlap with part of it). Perhaps split this into separate issues so the topics are more easily discussed/focused?

**Comment by erikerikson** [comment]
That sounds smart.  I'll do that.

**Comment by clemensv** [comment]
Can we close this one since it's been split into two issues?

**Comment by erikerikson** [comment]
Personally, I'd love all the concerns of persistence to be solved for me by the community.  I also believe that is a grammatically correct but meaningless statement.  Clemens: I think that you're right that taking those on would incur a lot of risk and distraction for the effort.

The reason I initially opened this was Doug's request for raising concerns and the concerns of this subject seemed absent from the spec.  It seems like a position that ought to be made explicit rather than implicit.  This would mean that each implementer would incur individual responsibility for those concerns and that correctness would be negotiated between service providers and their consumers.  The other two issues are merely a small subset of the concerns that fit under the concern of this issue and I created them to try and facilitate movement toward discussion and decision.

I suspect that probably it would be fairly simple to achieve some consensus in the community that persistence related issues should not be a focus of CloudEvents in the near term and that a formal statement of this could make it through the PR process.  I'd suggest that the acceptance of such a PR would answer the concerns of this issue and close it.  At the very least it would create the discussion this issue has failed to generate.

**Comment by duglin** [comment]
@erikerikson did you want to open a PR, perhaps for the primer, to let people know what our stance is w.r.t. persistence?

**Comment by erikerikson** [comment]
I will attempt to characterize what I understand our position to be.  Enjoy the weekend!
